File: fibonacci.c
# include <stdio.h>
# include <stdlib.h>
# include <time.h>
# include <inttypes.h>
#include <stdint.h>
typedef uint64_t ull;

void help() { // print usage instructions
    printf("Usage: ./fibonacci N TYPE\n");
    printf("TYPE: 1 = iterative, 2 = recursive, 3 = dynamic programming, 4 = all methods\n");
    printf("Example: ./fibonacci 25 4\n");
}

/**
 * Computes the nth Fibonacci number using a recursive approach.
 * Tracks the number of recursive calls as operations.
 * @param n    The index of the Fibonacci number to compute (n ≥ 0)
 * @param ops  Pointer to an operation counter (will be updated during recursion)
 * @return     The nth Fibonacci number
 */
ull fib_recursive(int n, ull* ops) {
    (*ops)++; // count each function call
    if (n == 0 || n == 1)  // base cases
        return n;
    return fib_recursive(n - 1, ops) + fib_recursive(n - 2, ops); // recursive case
}

/**
 * Computes the nth Fibonacci number using an iterative approach.
 * Tracks the number of addition operations performed.
 * @param n    The index of the Fibonacci number to compute (n ≥ 0)
 * @param ops  Pointer to an operation counter (will be updated during execution)
 * @return     The nth Fibonacci number
 */
ull fib_iterative(int n, ull* ops) {
    *ops = 0; // reset operation counter
    if (n == 0) return 0; // base case
    if (n == 1) return 1; // base case

    ull prev = 0; // stores F(n-2)
    ull curr = 1; // stores F(n-1)
    ull next; // will store F(n) in each iteration

    for (int i = 2; i <= n; i++) { // loop to compute F(2) thru F(n)
        next = prev + curr; // calculate next Fibonacci number: F(n) = F(n-1) + F(n-2)
        prev = curr; // shift previous values: F(n-2) becomes F(n-1)
        curr = next; // Update current value: F(n-1) becomes F(n)
        (*ops)++; // count each addition as one operation
    }
    return curr; // return the nth Fibonacci number
}

/**
 * Computes the nth Fibonacci number using bottom-up dynamic programming.
 * Stores intermediate results in a heap-allocated array and counts additions.
 * @param n    The index of the Fibonacci number to compute (n ≥ 0)
 * @param ops  Pointer to an operation counter (will be updated during execution)
 * @return     The nth Fibonacci number
 */
ull fib_dynamic(int n, ull* ops) {
    *ops = 0; // Reset operation counter
    if (n == 0) return 0; // base case
    if (n == 1) return 1; // base case

    ull* arr = (ull*)malloc((n + 1) * sizeof(ull)); // allocate space for Fibonacci values
    arr[0] = 0; // initialize base case F(0)
    arr[1] = 1; // initialize base case F(1)

    for (int i = 2; i <= n; i++) { // fill the array with Fibonacci values up to F(n)
        arr[i] = arr[i - 1] + arr[i - 2]; // F(n) = F(n-1) + F(n-2)
        (*ops)++; // count each addition
    }

    ull result = arr[n]; // store the result before freeing memory
    free(arr); // free allocated memory
    return result; // return the nth Fibonacci number
}


int main(int argc, char* argv[]) { 
    if (argc < 3) { // check for enough arguments
        printf("At least two arguments needed!\n");
        help(); // show usage instructions
        return 1; // exit if not enough arguments
    }

    // parse input arguments, convert to integers
    int n = atoi(argv[1]); 
    int type = atoi(argv[2]); 

    if (n < 0) { // check for negative input
        printf("n must be a non-negative integer.\n");
        return 1; // exit if n is negative
    }

    // declare variables for operation count, timing, result
    ull ops = 0;
    clock_t start, end;
    double runtime;
    ull result;

    switch (type) {
        case 1: // Iterative
            start = clock(); // start timing
            result = fib_iterative(n, &ops); // run Fibonacci iteratively
            end = clock(); // end timing
            runtime = (double)(end - start) / CLOCKS_PER_SEC; // calculate runtime
            printf("Iterative Fibonacci(%d) = %" PRIu64 "\n", n, result);
            printf("Time: %.6f, Ops: %" PRIu64 "\n", runtime, ops);
            break;

        case 2: // Recursive
            start = clock();
            result = fib_recursive(n, &ops);
            end = clock();
            runtime = (double)(end - start) / CLOCKS_PER_SEC;
            printf("Recursive Fibonacci(%d) = %" PRIu64 "\n", n, result);
            printf("Time: %.6f, Ops: %" PRIu64 "\n", runtime, ops);
            break;

        case 3: // Dynamic Programming
            start = clock();
            result = fib_dynamic(n, &ops);
            end = clock();
            runtime = (double)(end - start) / CLOCKS_PER_SEC;
            printf("Dynamic Programming Fibonacci(%d) = %" PRIu64 "\n", n, result);
            printf("Time: %.6f, Ops: %" PRIu64 "\n", runtime, ops);
            break;

        case 4: { // run all methods and print CSV-style output
            uint64_t ops_iter = 0, ops_dp = 0, ops_rec = 0;
            double runtime_iter, runtime_dp, runtime_rec;
            ull result;

            // Iterative
            clock_t start = clock();
            result = fib_iterative(n, &ops_iter);
            clock_t end = clock();
            runtime_iter = (double)(end - start) / CLOCKS_PER_SEC;

            // Dynamic Programming
            start = clock();
            result = fib_dynamic(n, &ops_dp);
            end = clock();
            runtime_dp = (double)(end - start) / CLOCKS_PER_SEC;

            // Recursive (only if n <= 40)
            if (n <= 40) {
                start = clock();
                result = fib_recursive(n, &ops_rec);
                end = clock();
                runtime_rec = (double)(end - start) / CLOCKS_PER_SEC;

                // print all results in CSV format, time and ops for each method
                printf("%.6f,%" PRIu64 ",%.6f,%" PRIu64 ",%.6f,%" PRIu64 "\n",
                    runtime_iter, ops_iter,
                    runtime_dp, ops_dp,
                    runtime_rec, ops_rec);
            } else {
                // skip recursive for large n, print placeholders
                printf("%.6f,%" PRIu64 ",%.6f,%" PRIu64 ",-,-\n",
                   runtime_iter, ops_iter,
                   runtime_dp, ops_dp);
                }
            break;
        }

        default: // invalid type
            printf("Type not supported.\n");
            help(); // show usage instructions
            return 1;

    }

}

File: fibonacci.py
import time

# Function: fib_iterative
# Description: Computes the nth Fibonacci number using iteration.
#              Tracks the number of addition operations performed.
# Parameters: n (int) - The index of the Fibonacci number to compute.
# Returns: tuple - (Fibonacci number, operation count)
def fib_iterative(n):
    ops = 0 # initialize operation counter
    if n == 0:
        return 0, ops # base case
    if n == 1:
        return 1, ops # base case

    prev, curr = 0, 1 # initialize first two Fibonacci numbers
    for _ in range(2, n + 1): # loop from 2 to n
        prev, curr = curr, prev + curr # update Fibonacci numbers
        ops += 1 # count each addition
    return curr, ops # return nth Fibonacci number and operation count


# Function: fib_recursive
# Description: Computes the nth Fibonacci number using recursion.
#              Uses a mutable list to track the number of recursive calls.
# Parameters: n (int) - The index of the Fibonacci number to compute.
#   ops_counter (list) - A single-element list used to count operations.
# Returns: int - The nth Fibonacci number
def fib_recursive(n, ops_counter):
    ops_counter[0] += 1 # count each recursive call
    if n == 0 or n == 1:
        return n # base cases
    # recursive calls for F(n-1) and F(n-2)
    return fib_recursive(n - 1, ops_counter) + fib_recursive(n - 2, ops_counter)


# Function: fib_dynamic
# Description: Computes the nth Fibonacci number using bottom-up dynamic programming.
#              Stores intermediate results in a list and counts additions.
# Parameters: n (int) - The index of the Fibonacci number to compute.
# Returns:
#   tuple - (Fibonacci number, operation count)
def fib_dynamic(n):
    ops = 0 # initialize operation counter
    if n == 0:
        return 0, ops # base case
    if n == 1:
        return 1, ops # base case

    arr = [0] * (n + 1) # create list to store Fibonacci numbers
    arr[0], arr[1] = 0, 1 # initialize base cases
    for i in range(2, n + 1): # loop from 2 to n
        arr[i] = arr[i - 1] + arr[i - 2] # compute Fibonacci number
        ops += 1 # count each addition
    return arr[n], ops # return nth Fibonacci number and operation count


# Function: run_and_time
# Description: Runs a Fibonacci function, times its execution, and prints results.
# Parameters: label (str) - Method name ("iterative", "recursive", or "dynamic")
#   func (function) - The Fibonacci function to run
#   n (int) - The index of the Fibonacci number to compute
def run_and_time(label, func, n):
    start = time.process_time() # start timing
    if label == "recursive":
        ops_counter = [0] # use list to track operations in recursion
        result = func(n, ops_counter)
        ops = ops_counter[0] # get operation count
    else:
        result, ops = func(n) # get result and operation count
    end = time.process_time() # end timing
    runtime = end - start # calculate runtime
    print(f"{label.capitalize()} Fibonacci({n}) = {result}") 
    print(f"Time: {runtime:.6f}, Ops: {ops}\n")

# Main runner
if __name__ == "__main__":
    import sys

    if len(sys.argv) < 2:
        print("Usage: python3 fibonacci_methods.py N")
        sys.exit(1)

    n = int(sys.argv[1])
    print(f"Running Fibonacci methods for n = {n}\n")

    run_and_time("iterative", fib_iterative, n)
    run_and_time("dynamic", fib_dynamic, n)

    if n <= 30:
        run_and_time("recursive", fib_recursive, n)
    else:
        print("Recursive method skipped for n > 30 due to performance.\n")

        